### PR TITLE
update command line documentation

### LINKED ISSUE #3020: Check documentation for "working with AiiDA/Command Line interface"
Check if it the section is updated to AiiDA 1.0, if it is correct, and if remove possible redundancies

https://aiida.readthedocs.io/projects/aiida-core/en/latest/working_with_aiida/index.html#

### PR DESCRIPTION
**Pull request by ltalirz**
fix #3020

 * hide unnecessary TOC (duplication of list)
 * reorder subsections, starting with parameters, then help strings
   (which contain info on parameters)
 * generally make text less wordy and move technical information to
   subsections
 * replace ID with PK

### CODE REVIEW

**Comment by ltalirz** [comment]
@sphuber ready for review

File: docs/source/verdi/verdi_user_guide.rst [hunk 1]
```diff
@@ -26,155 +48,108 @@ For example ``verdi process kill --help`` will display:
...
   * A list of the available options
 
-The ``[OPTIONS]`` and ``[PROCESSES]...`` tags in the usage description, denote the 'parameters' that the command takes.
-A more detailed description of the available options will be listed after the description of the commands functionality
-The positional arguments will only be described in the command description itself.
-For a more detailed explanation of the difference between options and arguments, see the section about command line :ref:`parameters<cli_parameters>`.
+The ``Usage:`` line encodes information on the command's parameters, e.g.:
 
+ * ``[OPTIONS]``: this command takes one (or more) options
+ * ``PROCESSES``: this command requires a process as a positional argument
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]: Since the importance here is the distinction of _requires_ and _optional_, maybe highlight these words?

File: docs/source/verdi/verdi_user_guide.rst [hunk 2]
```diff
@@ -26,155 +48,108 @@ For example ``verdi process kill --help`` will display:
...
-
-Parameters
-----------
-Most of the ``verdi`` commands and their subcommands can take one or multiple parameters.
-In the language of command line interfaces, these parameters come in two flavors:
-
-  * Options
-  * Arguments
+Multi-value options are followed by ``...`` in the help string and the ``Usage:`` line
+of the corresponding command will contains the 'endopts' marker. For example::
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]: Should be one line (or split at `[...] marker. For example::`)

File: docs/source/verdi/verdi_user_guide.rst [hunk 3]
```diff
@@ -26,155 +48,108 @@ For example ``verdi process kill --help`` will display:
...
 
+  Options:
+    -X, --codes CODE...             one or multiple codes identified by their
+                                    ID, UUID or label
+    -Y, --computers COMPUTER...     one or multiple computers identified by
+                                    their ID, UUID or label
+    -G, --groups GROUP...           one or multiple groups identified by their
+                                    ID, UUID or name
+    -N, --nodes NODE...             one or multiple nodes identified by their ID
+                                    or UUID
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]: There are more options here for `verdi export create`, could you show this perhaps by having a line like "`...`" or explicitly stating that this is a part of the full help response?

File: docs/source/verdi/verdi_user_guide.rst [hunk 4]
```diff
@@ -26,155 +48,108 @@ For example ``verdi process kill --help`` will display:
...
 
 .. _cli_profile:
 
 Profile
 -------
-AiiDA supports multiple profiles per installation, that can each be configured to use different databases.
-One of these profiles will always be marked as the default profile.
-To show the current default profile, execute the command::
+AiiDA supports multiple profiles per installation, one of which is marked as the default
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]: Take care to create new lines "properly" :)

File: docs/source/verdi/verdi_user_guide.rst [hunk 5]
```diff
@@ -26,155 +48,108 @@ For example ``verdi process kill --help`` will display:
...
-For this case ``verdi`` reserves a special character, the exclamation mark ``!`` that can be appended to the identifier.
-Before any type guessing is done, the command line will check for the presence of this marker, and if found will directly interpret the identifier as a ``LABEL``.
-For example, to solve ambiguity problems of the first two examples given in this section, one would have had to pass ``10!`` and ``deadbeef!``.
-The exclamation point would have forced them to be interpreted as a ``LABEL`` and ensured that the right group would be retrieved.
+ * "Primary Key" (PK): An integer, e.g. ``723``, that identifies your entity within your database (automatically assigned)
+ * `Universally Unique Identifier <https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)>`_ (UUID): A string, e.g. ``ce81c420-7751-48f6-af8e-eb7c6a30cec3`` that identifies your entity globally (automatically assigned)
+ * Label: A human-readable string, e.g. ``test_calculation`` (manually assigned)
 
-In summary, to guarantee correct identification of a specific type:
+Any ``verdi`` command that expects and identifier as a paramter will accept PKs, UUIDs and labels.
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
Any ``verdi`` command that expects an identifier as a parameter will accept PKs, UUIDs, and labels.
```

File: docs/source/verdi/verdi_user_guide.rst [hunk 6]
```diff
@@ -26,155 +48,108 @@ For example ``verdi process kill --help`` will display:
...
-The latter ambiguity, between an ``ID`` and ``UUID`` can always be resolved by passing a larger partial ``UUID``.
-Inevitably, eventually a non-numeric character or a dash will be included in the partial ``UUID``, rendering it an invalid ``ID`` and the identifier will be cast to the right type.
-The case of an identifier, that is intended to refer to a ``LABEL``, that just happens to also be a valid ``ID`` or ``UUID`` cannot be solved in this way.
-For this case ``verdi`` reserves a special character, the exclamation mark ``!`` that can be appended to the identifier.
-Before any type guessing is done, the command line will check for the presence of this marker, and if found will directly interpret the identifier as a ``LABEL``.
-For example, to solve ambiguity problems of the first two examples given in this section, one would have had to pass ``10!`` and ``deadbeef!``.
-The exclamation point would have forced them to be interpreted as a ``LABEL`` and ensured that the right group would be retrieved.
+ * "Primary Key" (PK): An integer, e.g. ``723``, that identifies your entity within your database (automatically assigned)
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]: This is great, however, could you afterwards (or here, but preferably afterwards) even more explicitly explain that PKs are local identifiers _only_ and the exact same database content may have completely different PKs for another user. And possibly add the now obvious advantage of UUIDs :)

File: docs/source/working_with_aiida/index.rst [hunk 7]
```diff
@@ -2,10 +2,16 @@
 Command line interface
 ======================
 
-The main way of interacting with AiiDA is through a command line interface tool called ``verdi``.
-Below you will find an overview of all the commands that are available with a link to a more detailed explanation of their usage and available parameters.
-But before you dive in, take a few minutes to read the :ref:`general concepts<cli_concepts>` that apply to the entire interface of ``verdi``.
-This will make understanding and using ``verdi`` a lot easier!
+One way of interacting with AiiDA is through the ``verdi`` command line interface.
```
Comment (CasperWA) [inline docs/source/working_with_aiida/index.rst]: Either here or in the other file, could you add the common acronym (CLI) after writing out command line interface? It is a kind favour to the reader, I think, but obviously not mandatory :)

File: docs/source/working_with_aiida/index.rst [hunk 8]
```diff
@@ -2,10 +2,16 @@
...
+
+Before checking out the individual commands below,
```
Comment (CasperWA) [inline docs/source/working_with_aiida/index.rst]: Take care of new lines

**Comment by CasperWA** [comment]
Hey @ltalirz, do you need help with this, or are you waiting for feedback from @sphuber? :)

**Comment by ltalirz** [comment]
just haven't come round to this yet, sorry (preparing the xiamen tutorial)

File: docs/source/verdi/verdi_user_guide.rst (hunk 6 above)
Comment (ltalirz) [reply docs/source/verdi/verdi_user_guide.rst]: thanks - I've put in a sentence on this

File: docs/source/working_with_aiida/index.rst (hunk 7 above)
Comment (ltalirz) [reply docs/source/working_with_aiida/index.rst]: done!

**Comment by ltalirz** [comment]
@CasperWA thanks for your suggestions, should be incorporated now!

File: docs/source/verdi/verdi_user_guide.rst [hunk 9]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
-.. _cli_parameters:
-
-Parameters
-----------
-Most of the ``verdi`` commands and their subcommands can take one or multiple parameters.
-In the language of command line interfaces, these parameters come in two flavors:
-
-  * Options
-  * Arguments
+Multi-value options are followed by ``...`` in the help string and the ``Usage:`` line of the corresponding command will contains the 'endopts' marker. For example::
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
Multi-value options are followed by ``...`` in the help string and the ``Usage:`` line of the corresponding command will contain the 'endopts' marker. For example::
```

File: docs/source/verdi/verdi_user_guide.rst [hunk 10]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
+    ...
 
 .. _cli_profile:
 
 Profile
 -------
-AiiDA supports multiple profiles per installation, that can each be configured to use different databases.
-One of these profiles will always be marked as the default profile.
-To show the current default profile, execute the command::
+AiiDA supports multiple profiles per installation, one of which is marked as the default and used unless another profile is requested
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
AiiDA supports multiple profiles per installation, one of which is marked as the default and used unless another profile is requested.
```

File: docs/source/verdi/verdi_user_guide.rst [hunk 11]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
-  * ``ID``: the integer primary key in the database
-  * ``UUID``: the universally unique identifier, a dash-separated hexadecimal string
-  * ``LABEL``: a custom string-based label
+When working with AiiDA entities, you need a way to *refer* to them on the command line.
+Any entity in AiiDA can be addressed via three identifiers:
 
-The ``ID`` and ``UUID`` identifiers follow the exact same rules for all the entities in AiiDA's data model.
-However, the ``LABEL`` will vary from entity to entity.
-For a ``Code`` instance it will be the ``label`` attribute, whereas for a ``Group`` instance, it will be its name.
+ * "Primary Key" (PK): An integer, e.g. ``723``, that identifies your entity within your database (automatically assigned)
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
 * "Primary Key" (PK): An integer, e.g. ``723``, identifying your entity within your database (automatically assigned)
```

File: docs/source/verdi/verdi_user_guide.rst [hunk 12]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
-
-===  =====================================  ========
-ID   UUID                                   LABEL
-===  =====================================  ========
-10   12dfb104-7b2b-4bca-adc0-1e4fd4ffcc88   group
-11   deadbeef-62ba-444f-976d-31d925dac557   10
-12   3df34a1e-5215-4e1a-b626-7f75b9586ef5   deadbeef
-===  =====================================  ========
+   PKs are easy to type and work as long as you stay within your database.
+   **When sharing data with others, however, always use UUIDs.**
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]: Excellent! Thanks :)

File: docs/source/verdi/verdi_user_guide.rst [hunk 13]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
-Before any type guessing is done, the command line will check for the presence of this marker, and if found will directly interpret the identifier as a ``LABEL``.
-For example, to solve ambiguity problems of the first two examples given in this section, one would have had to pass ``10!`` and ``deadbeef!``.
-The exclamation point would have forced them to be interpreted as a ``LABEL`` and ensured that the right group would be retrieved.
+In almost all cases, this will work out of the box.
+Since command line parameters are passed as strings, AiiDA needs to deduce the type of identifier from its content, which can fail in edge cases (see :ref:`cli_identifier_resolution` for details).
+You can take the following precautions in order to avoid such edge cases:
 
-In summary, to guarantee correct identification of a specific type:
+  * PK: no precautions needed
+  * UUID: no precautions needed for full UUIDs. Partial UUIDs should include at least one non-numeric character or dash.
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
  * UUID: no precautions needed for full UUIDs. Partial UUIDs should include at least one non-numeric character or dash
```

You don't have an ending period in any of your other lists (as far as I remember).

File: docs/source/verdi/verdi_user_guide.rst [hunk 14]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
--------------------
-The section on command line :ref:`parameters<cli_parameters>` explained that some commands support options and arguments that take one or multiple values.
-This is fairly typical for command line arguments, but slightly more unorthodox for options, that typically only ever take one value, or none at all if it is a flag.
-However, ``verdi`` has multiple commands where an option needs to be able to support options that take more than one value.
-Take for example the ``verdi export create`` command, with part of its help string::
+The logic for deducing the identifier type is as follows:
 
-  Usage: verdi export create [OPTIONS] [--] OUTPUT_FILE
+ 1. Try interpreting the identifier as a PK (integer)
+ 2. If this fails, try interpreting the identifier as a UUID (full or partial).
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
 2. If this fails, try interpreting the identifier as a UUID (full or partial)
```

Same as above.

File: docs/source/verdi/verdi_user_guide.rst [hunk 15]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
+12   3df34a1e-5215-4e1a-b626-7f75b9586ef5   deadbeef
+===  =====================================  ========
 
-Unfortunately, this leads to an ambiguity, as the 'greedy' multi value option ``-N`` will interpret the argument ``archive.aiida`` as an option value.
-This will cause the command to abort if the validation fails, but even worse it might be silently accepted.
-The root of the problem is that the multi value option needs to necessarily be greedy and cannot distinguish which value belongs to it and which value is just another argument.
-The typical solution for this problem is to use the so called 'endopts' marker, which is defined as two dashes ``--``, which can be used to mark the end of the options and clearly distinguish them from the arguments.
-Note that this is also indicated by the usage string of the command where it shows ``[--]`` between the ``[OPTIONS]`` and ``OUTPUT_FILE`` parameters, meaning that the ``--`` endopts marker can optionally be used.
-The previous command can therefore be made unambiguous as follows::
+ * trying to identify the first entity by its partial UUID ``12`` would match the third entity by its PK instead.
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
 * trying to identify the first entity by its partial UUID ``12`` would match the third entity by its PK instead
```

File: docs/source/verdi/verdi_user_guide.rst [hunk 16]
```diff
@@ -26,155 +48,112 @@ For example ``verdi process kill --help`` will display:
...
+ * trying to identify the second entity by its label ``10`` would match the first entity by its PK instead
+ * trying to identify the third entity by its label ``deadbeef`` would match the second entity on its partial UUID ``deadbeef`` instead.
```
Comment (CasperWA) [inline docs/source/verdi/verdi_user_guide.rst]:
```suggestion
 * trying to identify the third entity by its label ``deadbeef`` would match the second entity on its partial UUID ``deadbeef`` instead
```

**Comment by ltalirz** [comment]
Done :-)
I later realized I could have directly committed your suggestions - will do so next time

**Comment by ltalirz** [comment]
Does Sir @CasperWA approve?
